R/data.table.R

```diff
@@ -1885,67 +1891,165 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
            " but should be nrow(x)==", nrow(x))
   }
   if (!is.null(rownames)) {
-    # extract that column and drop it.
-    rownames.value = x[[rownames]]
+    # extract that column.
+    if (is.ff(x[[rownames]]))
+      rownames.value = x[[rownames]][] # bring the column into memory if ff object
+    else
+      rownames.value = x[[rownames]]
     dm = dim(x) - 0:1
     cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
   } else {
     dm = dim(x)
     cn = names(x)
-    X = x
+    rownames = ncol(x) + 1L # if no rownames to drop, set index to > ncol so that we can do [-rownames] rather than if (is.null(rownames)) {} else {} to drop the column in the rest of the code
   }
+  X = x
   if (any(dm == 0L))
     return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
+  p = length(cn)
   n = dm[1L]
   collabs = as.list(cn)
   class(X) = NULL
```

Comment:
Noting also that any type or class coercion of columns also creates new copies of at least those columns in memory, if not a copy of X itself. Will test further.

Reply:
Some profiling with pryr::object_size reveals why we create a copy `X = x` and strip its class. This does not make a full copy of `x`, but simply creates a list `X` whose elements are essentially pointers to the respective columns in `x`. Only once we modify that column (i.e. when coerced to another type/class) do we create a copy of the modified column in memory. This also means we can dramatically simplify the way we handle the `rownames` variable. We can simply drop this from `X` without modifying the input data.table `x` or creating a full new copy of `x` in memory.

Reply:
If we can move the column checks and type coercion into C we could avoid creating unnecessary copies of the coerced columns in memory, but this may be too onerous to code and maintain.